File: src/main/java/linq/SkipTake.java
package linq;

import linq.collections.enumerable.ArrayEnumerable;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class SkipTake {
    static final class ArraySkipTakeIterator<TSource> extends AbstractIterator<TSource> {
        private final ArrayEnumerable<TSource> source;
        private final int minIndexInclusive;
        private final int maxIndexInclusive;

        public ArraySkipTakeIterator(ArrayEnumerable<TSource> source, int minIndexInclusive, int maxIndexInclusive) {
            assert (source != null);
            assert (minIndexInclusive >= 0);
            assert (minIndexInclusive <= maxIndexInclusive);
            this.source = source;
            this.minIndexInclusive = minIndexInclusive;
            this.maxIndexInclusive = maxIndexInclusive;
        }

        @Override
        protected AbstractIterator<TSource> getClone() {
            return new ArraySkipTakeIterator<>(source, minIndexInclusive, maxIndexInclusive);
        }

        @Override
        public boolean moveNext() {
            // _state - 1 represents the zero-based index into the list.
            // Having a separate field for the index would be more readable. However, we save it
            // into _state with a bias to minimize field size of the iterator.
            var index = state - 1;
            if (index <= maxIndexInclusive - minIndexInclusive && index < source.count() - minIndexInclusive) {
                current = source.get(minIndexInclusive + index);
                ++state;
                return true;
            }

            close();
            return false;
        }

        @Override
        public <TResult> Enumerable<TResult> select(Function<TSource, TResult> selector) {
            return new Select.ArraySkipTakeSelectIterator<>(source, selector, minIndexInclusive, maxIndexInclusive);
        }

        @Override
        public AbstractIterator<TSource> skip(int count) {
            var minIndex = minIndexInclusive + count;
            return minIndex > maxIndexInclusive ? null : new ArraySkipTakeIterator<>(source, minIndex, maxIndexInclusive);
        }

        private static int getAdjustedCount(int minIndexInclusive, int maxIndexInclusive, int sourceCount) {
            if (sourceCount <= minIndexInclusive) {
                return 0;
            }

            return Math.min(sourceCount - 1, maxIndexInclusive) - minIndexInclusive + 1;
        }

        @Override
        public int getCount() {
            return getAdjustedCount(minIndexInclusive, maxIndexInclusive, source.count());
        }

        @Override
        public List<TSource> toList() {
            var count = getCount();

            var list = new ArrayList<TSource>(count);
            if (count != 0) {
                var sourceIndex = minIndexInclusive;
                for (var i = 0; i < count; i++, sourceIndex++) {
                    list.add(source.get(sourceIndex));
                }
            }

            return list;
        }
    }


    static final class EnumerableSkipTakeIterator<TSource> extends AbstractIterator<TSource> {
        private final Enumerable<TSource> source;
        private final int minIndexInclusive;
        private final int maxIndexInclusive; // -1 if we want everything past _minIndexInclusive.
        // If this is -1, it's impossible to set a limit on the count.
        private Enumerator<TSource> enumerator;

        public EnumerableSkipTakeIterator(Enumerable<TSource> source, int minIndexInclusive, int maxIndexInclusive) {
            assert (source != null);
            assert (minIndexInclusive >= 0);
            assert (maxIndexInclusive >= -1);
            // Note that although maxIndexInclusive can't grow, it can still be int.MaxValue.
            // We support partitioning enumerables with > 2B elements. For example, e.Skip(1).Take(int.MaxValue) should work.
            // But if it is int.MaxValue, then minIndexInclusive must != 0. Otherwise, our count may overflow.
            assert (maxIndexInclusive == -1 || (maxIndexInclusive - minIndexInclusive < Integer.MAX_VALUE));
            assert (maxIndexInclusive == -1 || minIndexInclusive <= maxIndexInclusive);

            this.source = source;
            this.minIndexInclusive = minIndexInclusive;
            this.maxIndexInclusive = maxIndexInclusive;
        }

        private boolean hasLimit() {
            return maxIndexInclusive != -1;
        }

        private int limit() {
            return maxIndexInclusive + 1 - minIndexInclusive;
        }

        @Override
        protected AbstractIterator<TSource> getClone() {
            return new EnumerableSkipTakeIterator<>(source, minIndexInclusive, maxIndexInclusive);
        }

        @Override
        public void close() {
            if (enumerator != null) {
                enumerator.close();
                enumerator = null;
            }

            super.close();
        }

        @Override
        public int getCount() {
            if (!hasLimit()) {
                return Math.max(source.count() - minIndexInclusive, 0);
            }

            try (var en = source.enumerator()) {
                var count = skipAndCount(maxIndexInclusive + 1, en);
                return Math.max(count - minIndexInclusive, 0);
            }
        }

        @Override
        public boolean moveNext() {
            int taken = state - 3;
            if (taken < -2) {
                close();
                return false;
            }

            switch (state) {
                case 1:
                    enumerator = source.enumerator();
                    state = 2;
                case 2:
                    assert (enumerator != null);
                    if (!skipBeforeFirst(enumerator)) {
                        // Reached the end before we finished skipping.
                        break;
                    }

                    state = 3;
                default:
                    assert (enumerator != null);
                    if ((!hasLimit() || taken < limit()) && enumerator.moveNext()) {
                        if (hasLimit()) {
                            // If we are taking an unknown number of elements, it's important not to increment _state.
                            // _state - 3 may eventually end up overflowing & we'll hit the Dispose branch even though
                            // we haven't finished enumerating.
                            state++;
                        }
                        current = enumerator.current();
                        return true;
                    }

                    break;
            }

            close();
            return false;
        }

        @Override
        public List<TSource> toList() {
            try (var en = source.enumerator()) {
                if (skipBeforeFirst(en) && en.moveNext()) {
                    int remaining = limit() - 1; // Max number of items left, not counting the current element.
                    int comparand = hasLimit() ? 0 : Integer.MIN_VALUE; // If we don't have an upper bound, have the comparison always return true.

                    var result = new ArrayList<TSource>();
                    do {
                        remaining--;
                        result.add(en.current());
                    }
                    while (remaining >= comparand && en.moveNext());

                    return result;
                }
            }

            return new ArrayList<>();
        }

        private boolean skipBeforeFirst(Enumerator<TSource> en) {
            return skipBefore(minIndexInclusive, en);
        }

        private static <TSource> boolean skipBefore(int index, Enumerator<TSource> en) {
            return skipAndCount(index, en) == index;
        }

        private static <TSource> int skipAndCount(int index, Enumerator<TSource> en) {
            assert (en != null);

            for (int i = 0; i < index; i++) {
                if (!en.moveNext()) {
                    return i;
                }
            }

            return index;
        }
    }
}

File: src/main/kotlin/linq/SkipTake.kt
package linq

import linq.collections.enumerable.ArrayEnumerable
import kotlin.math.max
import kotlin.math.min


class ArraySkipTakeIterator<TSource>(
    private val source: ArrayEnumerable<TSource>,
    private val minIndexInclusive: Int,
    private val maxIndexInclusive: Int
) : AbstractIterator<TSource>() {
    init {
        assert(minIndexInclusive >= 0)
        assert(minIndexInclusive <= maxIndexInclusive)
    }

    override fun clone(): AbstractIterator<TSource> =
        ArraySkipTakeIterator(source, minIndexInclusive, maxIndexInclusive)

    override fun moveNext(): Boolean {
        val index = state - 1
        if (index <= maxIndexInclusive - minIndexInclusive && index < source.size - minIndexInclusive) {
            current = source[minIndexInclusive + index]
            ++state
            return true
        }

        close()
        return false
    }

    override fun <TResult> itSelect(selector: (TSource) -> TResult): Enumerable<TResult> =
        ArraySkipTakeSelectIterator(source, selector, minIndexInclusive, maxIndexInclusive)

    override fun itSkip(count: Int): AbstractIterator<TSource>? {
        val minIndex = minIndexInclusive + count
        return if (minIndex > maxIndexInclusive) {
            null
        } else {
            ArraySkipTakeIterator(source, minIndex, maxIndexInclusive)
        }
    }

    override fun tryGetFirst(): TSource? {
        if (source.size > minIndexInclusive) {
            return source[minIndexInclusive]
        }
        return null
    }

    private fun getAdjustedCount(minIndexInclusive: Int, maxIndexInclusive: Int, sourceCount: Int): Int {
        if (sourceCount <= minIndexInclusive) {
            return 0
        }

        return min(sourceCount - 1, maxIndexInclusive) - minIndexInclusive + 1
    }

    override fun itCount(): Int = getAdjustedCount(minIndexInclusive, maxIndexInclusive, source.size)

    override fun getList(): List<TSource> {
        val count = itCount()

        val list = arrayListOf<TSource>()
        if (count != 0) {
            var sourceIndex = minIndexInclusive
            repeat(count) {
                list.add(source[sourceIndex])
                sourceIndex++
            }
        }

        return list
    }
}

class EnumerableSkipTakeIterator<TSource>(
    private val source: Enumerable<TSource>,
    private val minIndexInclusive: Int,
    private val maxIndexInclusive: Int
) : AbstractIterator<TSource>() {
    private var enumerator: Enumerator<TSource>? = null

    init {
        assert(minIndexInclusive >= 0)
        assert(maxIndexInclusive >= -1)
        assert(maxIndexInclusive == -1 || (maxIndexInclusive - minIndexInclusive < Integer.MAX_VALUE))
        { "will overflow!" }
        assert(maxIndexInclusive == -1 || minIndexInclusive <= maxIndexInclusive)
    }

    private val hasLimit: Boolean get() = maxIndexInclusive != -1

    private val limit: Int get() = maxIndexInclusive + 1 - minIndexInclusive

    override fun clone(): AbstractIterator<TSource> =
        EnumerableSkipTakeIterator(source, minIndexInclusive, maxIndexInclusive)

    override fun close() {
        enumerator?.close()
        enumerator = null
        super.close()
    }

    override fun itCount(): Int {
        if (!hasLimit) {
            return max(source.count() - minIndexInclusive, 0)
        }

        source.enumerator().use {
            val count = skipAndCount(maxIndexInclusive.toUInt() + 1u, it)
            assert(count != Int.MAX_VALUE.toUInt() + 1u || minIndexInclusive > 0) { "Our return value will be incorrect." }
            return max(count.toInt() - minIndexInclusive, 0)
        }
    }

    override fun moveNext(): Boolean {
        val taken = state - 3
        if (taken < -2) {
            close()
            return false
        }

        loop@ while (true) {
            when (state) {
                1 -> {
                    enumerator = source.enumerator()
                    state = 2
                    continue@loop
                }

                2 -> {
                    val enumerator = enumerator
                    enumerator.assertNotNull()

                    if (!skipBeforeFirst(enumerator)) {
                        break@loop
                    }

                    state = 3
                    continue@loop
                }

                else -> {
                    val enumerator = enumerator
                    enumerator.assertNotNull()
                    if ((!hasLimit || taken < limit) && enumerator.moveNext()) {
                        if (hasLimit) {
                            state++
                        }
                        current = enumerator.current
                        return true
                    }

                    break
                }
            }
        }

        close()
        return false
    }

    override fun itSkip(count: Int): AbstractIterator<TSource>? {
        val minIndex = minIndexInclusive + count

        if (!hasLimit) {
            if (minIndex < 0) {
                return EnumerableSkipTakeIterator(this, count, -1)
            }
        } else if (minIndex.toUInt() > maxIndexInclusive.toUInt()) {
            return null
        }

        assert(minIndex >= 0) { "We should have taken care of all cases when minIndex overflows." }
        return EnumerableSkipTakeIterator(source, minIndex, maxIndexInclusive)
    }

    override fun tryGetFirst(): TSource? {
        assert(!hasLimit || limit > 0)

        if (source is AbstractIterator<TSource>) {
            return source.tryGetElementAt(minIndexInclusive)
        }

        source.enumerator().use {
            if (skipBeforeFirst(it) && it.moveNext()) {
                return it.current
            }
        }

        return null
    }

    override fun getList(): List<TSource> {
        source.enumerator().use {
            if (skipBeforeFirst(it) && it.moveNext()) {
                var remaining = limit - 1
                val comparand = if (hasLimit) 0 else Int.MIN_VALUE

                val result = arrayListOf<TSource>()
                do {
                    remaining--
                    result.add(it.current)
                } while (remaining >= comparand && it.moveNext())

                return result
            }
        }

        return arrayListOf()
    }

    private fun skipBeforeFirst(en: Enumerator<TSource>): Boolean {
        return skipBefore(minIndexInclusive, en)
    }

    private fun skipBefore(index: Int, en: Enumerator<TSource>): Boolean {
        return skipAndCount(index, en) == index
    }

    private fun <TSource> skipAndCount(index: Int, en: Enumerator<TSource>): Int {
        assert(index >= 0)
        return skipAndCount(index.toUInt(), en).toInt()
    }

    private fun <TSource> skipAndCount(index: UInt, en: Enumerator<TSource>): UInt {
        for (i in 0u until index) {
            if (!en.moveNext()) {
                return i
            }
        }

        return index
    }
}